mlflow/pyfunc/utils/data_validation.py

```diff
@@ -0,0 +1,98 @@
+import warnings
+from functools import lru_cache, wraps
+
+from mlflow.models.signature import _extract_type_hints, _is_context_in_predict_function_signature
+from mlflow.types.type_hints import (
+    _convert_data_to_type_hint,
+    _infer_schema_from_type_hint,
+    _signature_cannot_be_inferred_from_type_hint,
+    _validate_example_against_type_hint,
+)
+from mlflow.utils.annotations import filter_user_warnings_once
+
+
+def pyfunc(func):
+    """
+    A decorator that forces data validation against type hint of the input data
+    in the wrapped method. It is no-op if the type hint is not supported by MLflow.
+
+    .. note::
+        The function that applies this decorator must be a valid `predict` function
+        of `mlflow.pyfunc.PythonModel`, or a callable that takes a single input.
+    """
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        if type_hint := _predict_type_hint_check(func):
```

Comment:
Can we do this outside the wrapper so that users can see the unsupported type hint nessage when defining the function or pyfunc class? In current implementation, users need to run a prediction at least once.

Reply:
Seems this works if @pyfunc is explicitly used (on callables), but in PythonModel we're using pyfunc(...) inside __getattribute__ so the warning will only be shown if predict method is called 🤔

Reply:
@serena-ruan I think we can use `__init_subclass__` to validate if the `predict` function of any subclass when the class is defined (and maybe also extract model signature at that timing).

Reply:
Updated!